lib/ansible/modules/windows/win_xml.ps1

```diff
@@ -0,0 +1,241 @@
+	Copy-Item $path $backuppath;
+	return $backuppath;
+}
+
+$params = Parse-Args $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+$debug_level = Get-AnsibleParam -obj $params -name "_ansible_verbosity" -type "int"
+$debug = $debug_level -gt 2
+
+$dest = Get-AnsibleParam $params "path" -type "path" -FailIfEmpty $true -aliases "dest", "file"
+$fragment = Get-AnsibleParam $params "fragment" -type "str" -FailIfEmpty $true -aliases "xmlstring"
+$xpath = Get-AnsibleParam $params "xpath" -type "str" -FailIfEmpty $true
+$backup = Get-AnsibleParam $params "backup" -type "bool" -Default $false
+$type = Get-AnsibleParam $params "type" -type "str" -Default "element" -ValidateSet "element", "attribute", "text"
+$attribute = Get-AnsibleParam $params "attribute" -type "str" -FailIfEmpty ($type -eq "attribute")
+$state = Get-AnsibleParam $params "state" -type "str" -Default "present"
+
+$result = @{
+    changed = $false
+}
+
+If (-Not (Test-Path -Path $dest -PathType Leaf)){
+    Fail-Json $result "Specified path $dest does not exist or is not a file."
+}
+
+[xml]$xmlorig = $null
+Try {
+    [xml]$xmlorig = Get-Content -Path $dest
+}
+Catch {
+    Fail-Json $result "Failed to parse file at '$dest' as an XML document: $($_.Exception.Message)"
+}
+
+$namespaceMgr = New-Object System.Xml.XmlNamespaceManager $xmlorig.NameTable
+$namespace = $xmlorig.DocumentElement.NamespaceURI
+$localname = $xmlorig.DocumentElement.LocalName
+
+$namespaceMgr.AddNamespace($xmlorig.$localname.SchemaInfo.Prefix, $namespace)
+
+if ($type -eq "element") {
+    $xmlchild = $null
+    Try {
+        $xmlchild = [xml]$fragment
+    } Catch {
+        Fail-Json $result "Failed to parse fragment as XML: $($_.Exception.Message)"
+    }
+
+    $child = $xmlorig.CreateElement($xmlchild.get_DocumentElement().get_Name(), $xmlorig.get_DocumentElement().get_NamespaceURI())
+    Copy-Xml $child $xmlchild.DocumentElement $xmlorig
+
+    $node = $xmlorig.SelectSingleNode($xpath, $namespaceMgr)
+    if ($node.get_NodeType() -eq "Document") {
+        $node = $node.get_DocumentElement()
+    }
+    $elements = $node.get_ChildNodes()
+    [bool]$present = $false
+    [bool]$changed = $false
+    if ($elements.get_Count()) {
+        if ($debug) {
+            $err = @()
+            $result.err = {$err}.Invoke()
+        }
+        foreach ($element in $elements) {
+            try {
+                Compare-XmlDocs $child $element
+                $present = $true
+                break
+            } catch {
+                if ($debug) {
+                    $result.err.Add($_.Exception.ToString())
+                }
+            }
+        }
+        if (!$present -and ($state -eq "present")) {
+            [void]$node.AppendChild($child)
+            $result.msg = "xml added"
+            $changed = $true
+        } elseif ($present -and ($state -eq "absent")) {
```

Comment:
I am trying to remove an element from xml file but its not working for me. I am trying following, Please correct me if i am doing wrong.

`- name: Running XML Transformations with state Absent
     win_xml:
       path:      "path\to\xml\file"
       xpath:     "/configuration/configSections/sectionGroup/section[@name='authentication']"
       state:     "absent"
       fragment: " "`

Reply:
Should the xpath be `//configuration/configSections/sectionGroup/section[@name='authentication']` - double initial slash?

Reply:
@petemounce it worked for other paramerters without adding double initial slash. its not working only if i am trying to delete an element.

Reply:
I dont think you can have an empty fragment. The xpath should find the root of your XML where the child fragment is, then it will remove that child fragment.

Reply:
Ah, pardon.

Reply:
Could you provide the XML or a facsimile thereof? I I have enough info to reproduce and test what you are trying to do but need some XML to test with.

Reply:
<?xml version="1.0" encoding="utf-8"?>
<!--
  For more information on how to configure your ASP.NET application, please visit
  http://go.microsoft.com/fwlink/?LinkId=169433
  -->
<configuration>
  <configSections>
      <sectionGroup name="applicationSettings" type="System.Configuration.ApplicationSettingsGroup, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" >
          <section name="GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings" type="System.Configuration.ClientSettingsSection, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" requirePermission="false" />
      </sectionGroup>
	 <sectionGroup name="GKE.ds.security" type="GKE.DS.Security.Configuration.SecuritySectionGroup, GKE.DS.Security">
      <section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />
      <section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />    
     </sectionGroup>
	<!--
  Adding a multiline comment
  second line comment
  third line
  -->
  </configSections>
  <system.web>
    <compilation debug="true" targetFramework="4.5" />
    <httpRuntime targetFramework="4.5" />
  </system.web>
  <appSettings>
  </appSettings>
<applicationSettings>
    <GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
      <setting name="Setting1" serializeAs="String">
        <value>Setting1Value</value>
      </setting>
    </GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
  </applicationSettings>
</configuration>

Reply:
We need to remove these lines.
`<section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />
      <section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />    `

Reply:
I apologize, I was confused. If you were adding an element you would specify the XPath and the entire element you want added for the fragment. I think for removing you would also have to specify the entire fragment you want removed. The module recursively checks for "equality" so you  have to have the entire fragment. Also fragments are not treated as XPath, sorry for that mistake. 

So your xpath would be "/configuration/configSections" (n.b. no ending slash) and you would need to use the module twice, once with fragment as '<section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />' and again with fragment as '<section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />'

Reply:
Ok, I have tested this and verified that this works. The way GitHub formats the code blocks makes it hard to read the XML you're working with. The XPath should be `'/configuration/configSections/sectionGroup[@name="GKE.ds.security"]'` and the full playbook used is as follows:

`
---
 - name: win_xml module test with absent
   hosts: []
   tasks:
     - name: run the win_xml module
       win_xml:
          path: C:\\Users\richardcs\Temp\test.xml
          xpath: '/configuration/configSections/sectionGroup[@name="GKE.ds.security"]'
          fragment: '<section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />'
          type: element
          state: absent

Reply:
Okay I will give it a try.

Reply:
@richardcs its worked ..!! thanks for helping me out. In Fragment , you have to give element exactly the same as in your xml file.

Reply:
@abhijeetka yes, the logic there is if you were adding it, the default behavior, you would have to give the entire XML fragment you wanted to add, obviously. So to delete, you do the same. Glad it worked for you.

Reply:
@richardcs one last thing that i want to know, Can we add the capability to delete even the attributes from the elements?
We are trying to remove only the attributes from the xml element.
For. Ex <language default="eng" style="arial"></language>
We want to delete the default tag only.

Reply:
I don't think the current module supports absent for attributes. If you give me the .xml you are working with I can probably add it.

Reply:
Here is the xml file.
```
<?xml version="1.0" encoding="utf-8"?>
<!--
  For more information on how to configure your ASP.NET application, please visit
  http://go.microsoft.com/fwlink/?LinkId=169433
  -->
<configuration>
  <configSections>
      <sectionGroup name="applicationSettings" type="System.Configuration.ApplicationSettingsGroup, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" >
          <section name="GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings" type="System.Configuration.ClientSettingsSection, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" requirePermission="false" />
      </sectionGroup>
	 <sectionGroup name="GKE.ds.security" type="GKE.DS.Security.Configuration.SecuritySectionGroup, GKE.DS.Security">
      <section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />
      <section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />    
     </sectionGroup>
	<!--
  Adding a multiline comment
  second line comment
  third line
  -->
  </configSections>
  <system.web>
    <compilation debug="true" targetFramework="4.5" />
    <httpRuntime targetFramework="4.5" />
  </system.web>
  <appSettings>
  </appSettings>
<applicationSettings>
    <GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
      <setting name="Setting1" serializeAs="String">
        <value>Setting1Value</value>
      </setting>
    </GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
  </applicationSettings>
</configuration>
```
we want to remove a attribute `serializeAs="String"` which is present in  `<setting name="Setting1" serializeAs="String">` element.

Reply:
@abhijeetka I modified the win_xml module to support your use case. It was tested with this playbook. 
```
---
 - name: win_xml module test with absent attribute
   hosts: []
   tasks:
     - name: run the win_xml module
       win_xml:
          path: C:\\Users\richardcs\Temp\test.xml
          xpath: '/configuration/applicationSettings/GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings/setting'
          type: attribute
          fragment: String
          attribute: serializeAs
          state: absent
```

Reply:
@richardcs  that works like a charm ...!! Thanks for adding this functionality.